app/vminsert/netstorage/insert_ctx.go

```diff
@@ -165,9 +189,24 @@ func (ctx *InsertCtx) GetStorageNodeIdx(at *auth.Token, labels []prompb.Label) i
 	}
 	h := xxhash.Sum64(buf)
 	ctx.labelsBuf = buf
+	for rgID, rg := range replicationGroups {
+		if len(rg.storageNodes) == 1 {
+			// Fast path - only a single storage node.
+			groupPairs = append(groupPairs, StorageNodeGroupID{Group: rgID, Idx: 0})
+			continue
+		}
+
+		idx := int(jump.Hash(h, int32(len(rg.storageNodes))))
+		groupPairs = append(groupPairs, StorageNodeGroupID{Group: rgID, Idx: idx})
+	}
+	return groupPairs
+}
+
+// AllReplicationGroupsFailed - check if all replication groups have failed
+func (ctx *InsertCtx) AllReplicationGroupsFailed(replicationGroup string) bool {
```

Comment:
This function isn't needed if counting errors in the loop over `replicationGroups` as described above